### PR TITLE
etna: add ShaderProgram manager

### PR DESCRIPTION
**Pull request by FptrP**
- Added shader program manager with automatic
descriptor set layout and pipeline layout generation

### CODE REVIEW

File: etna/include/etna/ShaderProgram.hpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    void reload(vk::Device device);
+    void reset(vk::Device device);
+
+    const auto &getResources() const { return resources; }
+    vk::ShaderModule getVkModule() const { return vkModule; }
+    vk::ShaderStageFlagBits getStage() const { return stage; }
+    const std::string &getName() const { return entryPoint; }
+    vk::PushConstantRange getPushConst() const { return pushConst; }
+
+    bool getPushConstUsed() const { return pushConst.size !=0u; }
```
Comment (AlexandrShcherbakov) [inline etna/include/etna/ShaderProgram.hpp:90]: getPushConstCount/getPushConstSize maybe?

File: etna/include/etna/ShaderProgram.hpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    std::unique_ptr<ShaderProgramInternal> createProgram(std::vector<uint32_t> &&modules);
+    void destroyProgram(std::unique_ptr<ShaderProgramInternal> &ptr);
+    //void reloadPrograms();
+
+    vk::Device getDevice() const { return vkDevice; } /*make device global*/
+  };
+}
+
+#endif
```
Comment (AlexandrShcherbakov) [inline etna/include/etna/ShaderProgram.hpp:166]: Add an empty line.

File: etna/include/etna/ShaderProgram.hpp [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    
+      vk::PushConstantRange pushConst {};
+      vk::PipelineLayout progLayout {nullptr};
+    };
+
+    std::unordered_map<std::string, uint32_t> programNames;
+    std::vector<std::unique_ptr<ShaderProgramInternal>> programs;
+
+    std::unique_ptr<ShaderProgramInternal> createProgram(std::vector<uint32_t> &&modules);
+    void destroyProgram(std::unique_ptr<ShaderProgramInternal> &ptr);
```
Comment (AlexandrShcherbakov) [inline etna/include/etna/ShaderProgram.hpp:159]: Why this destroy call is not a part of ShaderProgramInternal destructor?

File: etna/include/etna/ShaderProgram.hpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    DescriptorSetLayoutCache(const DescriptorSetLayoutCache&) = delete;
+    DescriptorSetLayoutCache &operator=(const DescriptorSetLayoutCache&) = delete; 
+
+  private:
+    std::unordered_map<DescriptorSetInfo, DescriptorLayoutId, DescriptorSetLayoutHash> map;
+    std::vector<DescriptorSetInfo> descriptors;
+    std::vector<vk::DescriptorSetLayout> vkLayouts;
+  };
+
+  struct ShaderModule
```
Comment (AlexandrShcherbakov) [inline etna/include/etna/ShaderProgram.hpp:77]: Move ShaderModule and ShaderProgramManager in another header, please.

File: etna/include/etna/ShaderProgram.hpp [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+#include <array>
+#include <bitset>
+#include <vector>
+#include <unordered_map>
+#include <memory>
+
+namespace etna
+{
+  constexpr uint32_t MAX_PROGRAM_DESCRIPTORS = 4u;
+  constexpr uint32_t MAX_DESCRIPTOR_BINDINGS = 8u; /*If you are out of bindings, try using arrays of images/samplers*/
```
Comment (AlexandrShcherbakov) [inline etna/include/etna/ShaderProgram.hpp:18]: I think, 32 is better here.

File: etna/source/ShaderProgram.cpp [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+    if (binding.descriptorType == vk::DescriptorType::eUniformBufferDynamic || binding.descriptorType == vk::DescriptorType::eStorageBufferDynamic)
+      dynOffsets++;
+  }
+  
+  void DescriptorSetInfo::clear()
+  {
+    maxUsedBinding = 0;
+    dynOffsets = 0;
+    usedBindings.reset();
+    for (auto &binding : bindings)
```
Comment (AlexandrShcherbakov) [inline etna/source/ShaderProgram.cpp:40]: Nit. std::fill.

File: etna/source/ShaderProgram.cpp [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+      binding = vk::DescriptorSetLayoutBinding {};
+  }
+
+  void DescriptorSetInfo::parseShader(vk::ShaderStageFlagBits stage, const SpvReflectDescriptorSet &spv)
+  {
+    for (uint32_t i = 0u; i < spv.binding_count; i++)
+    {
+      const auto &spvBinding = *spv.bindings[i];
```
Comment (AlexandrShcherbakov) [inline etna/source/ShaderProgram.cpp:48]: What if not all bindings are used in shader?

File: etna/source/ShaderProgram.cpp [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+
+  #define SPVR_ASSER(res) if ((res) != SPV_REFLECT_RESULT_SUCCESS) throw std::runtime_error {"SPVReflect error"}
+
+  void ShaderModule::reload(vk::Device device)
+  {
+    reset(device);
+
+    auto code = read_file(path);
+    vk::ShaderModuleCreateInfo info {};
+    info.setPCode(reinterpret_cast<const uint32_t*>(code.data()));
```
Comment (AlexandrShcherbakov) [inline etna/source/ShaderProgram.cpp:197]: if (code.size() % 4) throw something.

File: etna/source/ShaderProgram.cpp [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+  struct SpvModDeleter
+  {
+    SpvModDeleter() {}
+    void operator()(SpvReflectShaderModule *mod) const
+    {
+      spvReflectDestroyShaderModule(mod);
+    }
+  };
+
+  #define SPVR_ASSER(res) if ((res) != SPV_REFLECT_RESULT_SUCCESS) throw std::runtime_error {"SPVReflect error"}
```
Comment (AlexandrShcherbakov) [inline etna/source/ShaderProgram.cpp:189]: SPVR_ASSERT

**Review by AlexandrShcherbakov — CHANGES_REQUESTED** [review]
Let's split header and source files. Other comments could be fixed later.

File: etna/include/etna/ShaderProgram.hpp (hunk 1 above)
Comment (FptrP) [reply etna/include/etna/ShaderProgram.hpp]: Removed this function

File: etna/include/etna/ShaderProgram.hpp (hunk 2 above)
Comment (FptrP) [reply etna/include/etna/ShaderProgram.hpp]: Done

File: etna/include/etna/ShaderProgram.hpp (hunk 3 above)
Comment (FptrP) [reply etna/include/etna/ShaderProgram.hpp]: added destroy and reload functionality as methods

File: etna/include/etna/ShaderProgram.hpp (hunk 4 above)
Comment (FptrP) [reply etna/include/etna/ShaderProgram.hpp:17]: Moved DescriptorSetInfo and cache to separate file

File: etna/source/ShaderProgram.cpp (hunk 7 above)
Comment (FptrP) [reply etna/source/ShaderProgram.cpp]: Looks like this values are not nullptr

**Pull request merged**